README.md
# This project is maintained

Took over from previous owner because [the old repo](https://github.com/babariviere/flutter_sms) isn't maintained.

## Reason I took over

I am making an SMS app (that will be open source when it is more stable) and I need a good package for all the CRUD actions that I need to do with my messages.

# Flutter SMS

This is an SMS library for Flutter.

**It only supports Android for now.** iOS and MMS is not in the scope of this project.

## Extra

I have forked this repository so I can improve and add my own functionality.

## Getting Started

For help getting started with Flutter, view the online
[documentation](https://flutter.io/).

For help on editing plugin code, view the [documentation](https://flutter.io/platform-plugins/#edit-code).

## Installation and Usage

Once you're familiar with Flutter you may install this package adding `sms_maintained` (0.2.5 or higher) to the dependencies list of the `pubspec.yaml` file as follow:

```yaml
---
dependencies:
  flutter:
    sdk: flutter

  sms_maintained: ^0.2.5
```

Then run the command `flutter packages get` on the console.

## Querying SMS messages

Add the import statement for `sms_maintained` and create an instance of the _SmsQuery_ class:

```dart
import 'package:sms_maintained/sms.dart';

void main() {
  SmsQuery query = new SmsQuery();
}

```

## Getting all SMS messages

```dart
List<SmsMessage> messages = await query.getAllSms;
```

**Note**: _the use of `await` keyword means that `getAllSms` is resolved asynchronously and a Future is retorned. To use await, you must be in a function marked `async`._

## Filtering SMS messages

The method `querySms` from the `SmsQuery` class returns a list of filtered SMS messages that match the supplied parameters. For example, for querying all the SMS messages sent and received, you could this code:

```dart
await query.querySms({
    kinds: [SmsQueryKind.Inbox, SmsQueryKind.Sent]
});
```

You can also query all the SMS messages sent and received from a specific contact:

```dart
await query.querySms({
    address: getContactAddress()
});
```

## Getting all Threads Conversations

With `SmsQuery` you can also get the entire list of conversations:

```dart
List<SmsThread> threads = await query.getAllThreads;
```

## Getting the Contact info

Each conversation thread is related with a Contact.
The class `Contact` contains all the info of a thread contact (address, photo, full name).
To get access to `Contact` class you must import `'package:sms_maintained/contact.dart'` into your dart file:

```dart
import 'package:sms_maintained/contact.dart';

void main() {
  ...
  Contact contact = threads.first.contact;
  print(contact.address);
}
```

## Querying Contact

You can also query a contact by its address _(phone number)_:

```dart
import 'package:sms_maintained/contact.dart';

void main() {
  ContactQuery contacts = new ContactQuery();
  Contact contact = await contacts.queryContact(someAddress());
  print(contact.fullName);
}
String getAddress() { return someRandomAddress; }
```

## The Contact photo

You can retrieve the photo of the contact (full size or thumbnail):

```dart
...
Uint8List fullSize = contact.photo.bytes;
Uint8List thumbnail = contact.thumbnail.bytes;
```

## User Profile

Some times it is useful to request basic info of the phone owner, like the contact photo, addresses, etc.

```dart
import 'package:sms_maintained/contact.dart';

UserProfileProvider provider = new UserProfileProvider();
UserProfile profile = await provider.getUserProfile();
print(profile.fullName);
```

## Sending SMS

What about sending a SMS? All you have to do is to create an instance of the `SmsSender` class:

```dart
import 'package:sms_maintained/sms.dart';

void main() {
  SmsSender sender = new SmsSender();
  String address = someAddress();
  ...
  sender.sendSms(new SmsMessage(address, 'Hello flutter!'));
}
```

To be notified when the message is sent and/or delivered, you must add a listener to your message:

```dart
import 'package:sms_maintained/sms.dart';

void main() {
  SmsSender sender = new SmsSender();
  String address = someAddress();
  ...
  SmsMessage message = new SmsMessage(address, 'Hello flutter!');
  message.onStateChanged.listen((state) {
    if (state == SmsMessageState.Sent) {
      print("SMS is sent!");
    } else if (state == SmsMessageState.Delivered) {
      print("SMS is delivered!");
    }
  });
  sender.sendSms(message);
}
```

Some times it is useful to be notified of delivered messages regardless of the message. To do that you must subscribe to the `onSmsDelivered` of the `SmsSender` class instance:

```dart
void main() {
...
SmsSender sender = new SmsSender();
sender.onSmsDelivered.listen((SmsMessage message){
  print('${message.address} received your message.');
}));
}
```

You can also send with another SimCard:

```dart
void main() {
SimCardsProvider provider = new SimCardsProvider();
SimCard card = await provider.getSimCards()[0];
SmsSender sender = new SmsSender();
SmsMessage message = new SmsMessage("address", "message");
sender.sendSMS(message, simCard: card);
}
```

**Note**: Using the `onSmsDelivered` from the `SmsSender` will only notify to listeners of messages that has been sent through `SmsSender.send()`.

## Receiving SMS

If you want to be notified for incoming new messages you must subscribe to an instance of the `SmsReceiver` class:

```dart
import 'package:sms_maintained/sms.dart';

void main() {
  SmsReceiver receiver = new SmsReceiver();
  receiver.onSmsReceived.listen((SmsMessage msg) => print(msg.body));
}
```

## Deleting SMS

Only deleting one by one is available. Don't forget to make your sms app the default sms app.
[blogpost guide](https://android-developers.googleblog.com/2013/10/getting-your-sms-apps-ready-for-kitkat.html)

```dart
import 'package:sms_maintained/sms.dart';

void main() {
SmsRemover smsRemover = SmsRemover();
<boolean value> = await smsRemover.removeSms(fromAddress);
}
```

## Roadmap

- [x] SMS Receiver
- [x] SMS Sender
- [x] SMS Delivery
- [x] SMS Query
- [x] SMS Thread
- [x] Contact
- [x] Contact Photo (full size, thumbnail)
- [x] User profile (basic info)
- [x] SMS Delete One-by-one
- [ ] Improve speed.
- [ ] Add a caching layer.
- [ ] Add a helper.
- [ ] Add observables so you don't need to manually poll for changes.
- [ ] ...

MMS and iOS is not in the scope of this project. If someone wants to add the code and make a merge request for it, I am happy to include it.

If there are requests you can always make an issue. I will see what I can do.

## Contributors

- babariviere [github](https://github.com/babariviere)
- joanpablo [github](https://github.com/joanpablo)
- saifulfrank [github](https://github.com/saifulfrank)
- geordyvc [github](https://github.com/geordyvcErasmus)

### App logo

Designed and created by [Imrul kayes](https://github.com/saifulfrank)

## Contributions

Any contribution is welcome.

## Donations

I'm a student so money isn't pouring in. Feel free to make a donation.
[PayPal](https://www.paypal.me/GeordyVC)


CHANGELOG.md
## [0.2.6] - 2019-11-11

* Allow multipart sms sending

## [0.2.5] - 2019-03-17

* Add way to remove one message at a time.
* Took over from the previous author because he doesn't maintain the project anymore.

## [0.2.4] - 2018-09-30

* Add fail state

## [0.2.3] - 2018-09-29

* Fix issue with SMS state handler

## [0.2.2] - 2018-09-05

* Update dart version

## [0.2.1] - 2018-08-13

* Fix issue with `toString` on error

## [0.2.0] - 2018-08-09

* Multi SIM card support

## [0.1.6] - 2018-08-07

* Fix permission error when sending SMS on Android O and above

## [0.1.5] - 2018-08-06

* Fix permission error on Android M and above

## [0.1.4] - 2018-06-16

* Error on contact lookup because of unknown size fixed

## [0.1.3] - 2018-05-29
* Error in Android 4.x because of missing permission 'READ_PROFILE' fixed
* Updated Flutter environment sdk to ">=1.19.0 <2.0.0"
* Example app code refactoring

## [0.1.2] - 2018-05-26
* Photo and thumbnail of a contact get loaded with the contact, no need to request them later.
* Example app UI slightly enhanced.

## [0.1.1] - 2018-05-14

* Sms Delivery with Dart Streams and Event Channels
* Fixed some minor errors
* Fixed error on User Profile. The method 'getUserProfile()' will always return an instance of UserProfile even when no user profile configured in the device.
* Fixed error in SmsThread.findContact when first message is a draft.

## [0.1.0] - 2018-04-23

* Sms Delivery

## [0.0.10] - 2018-04-18

* Automatically assign contact to created SmsThread
* Small fix for 'date' and 'dateSent' of SmsMessage
* Fixed error on multiple permissions requests
* Get basic User Profile info.
* Fix gradle build

## [0.0.9] - 2018-03-15

* backward compatibility for Contact (pull request #1 by joanpablo)

## [0.0.8] - 2018-03-10

* SmsReceiver creates all needed field
* Add basic contact support

## [0.0.7] - 2018-03-09

* Small fix for SMS thread

## [0.0.6] - 2018-03-09

* Implementation of SMS thread

## [0.0.5] - 2018-03-09

* Support for dart < 2.0.0

## [0.0.4] - 2018-03-09

* Better query for SMS, now you can query by multiple kind instead of only one.

## [0.0.3] - 2018-03-09

* Better handling of permissions
* Implementation of SmsQuery

## [0.0.2] - 2018-03-08

* Implementation of SmsSender

## [0.0.1] - 2018-03-08

* Implementation of SmsReceiver.


android/src/main/java/com/babariviere/sms/SmsSender.java
package com.babariviere.sms;

import android.Manifest;
import android.annotation.TargetApi;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.provider.Telephony;
import android.telephony.SmsManager;
import android.util.Log;

import com.babariviere.sms.permisions.Permissions;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.UUID;
import java.util.ArrayList;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;

import static io.flutter.plugin.common.PluginRegistry.Registrar;
import static io.flutter.plugin.common.PluginRegistry.RequestPermissionsResultListener;

/**
 * Created by babariviere on 08/03/18.
 */

@TargetApi(Build.VERSION_CODES.DONUT)
class SmsSenderMethodHandler implements RequestPermissionsResultListener {
    private static final SmsManager sms = SmsManager.getDefault();
    private final String[] permissionsList = new String[]{Manifest.permission.SEND_SMS, Manifest.permission.READ_PHONE_STATE};
    private MethodChannel.Result result;
    private String address;
    private String body;
    private int sentId;
    private Integer subId;
    private final Registrar registrar;

    SmsSenderMethodHandler(Registrar registrar, MethodChannel.Result result, String address, String body, int sentId, Integer subId) {
        this.registrar = registrar;
        this.result = result;
        this.address = address;
        this.body = body;
        this.sentId = sentId;
        this.subId = subId;
    }

    void handle(Permissions permissions) {
        if (permissions.checkAndRequestPermission(permissionsList, Permissions.SEND_SMS_ID_REQ)) {
            sendSmsMessage();
        }
    }

    @Override
    public boolean onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode != Permissions.SEND_SMS_ID_REQ) {
            return false;
        }
        boolean isOk = true;
        for (int res : grantResults) {
            if (res != PackageManager.PERMISSION_GRANTED) {
                isOk = false;
                break;
            }
        }
        if (isOk) {
            sendSmsMessage();
            return true;
        }
        result.error("#01", "permission denied for sending sms", null);
        return false;
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    private void sendSmsMessage() {
        SmsManager sms;
        if (this.subId == null) {
            sms = SmsManager.getDefault();
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
                sms = SmsManager.getSmsManagerForSubscriptionId(this.subId);
            } else {
                result.error("#03", "this version of android does not support multicard SIM", null);
                return;
            }
        }
        ArrayList<String> parts = sms.divideMessage(body);
        ArrayList<PendingIntent> sentIntents = new ArrayList();
        ArrayList<PendingIntent> deliveredIntents = new ArrayList();

        for (String part : parts) {
            Intent sentIntent = new Intent("SMS_SENT");
            sentIntent.putExtra("sentId", sentId);
            PendingIntent sentPendingIntent = PendingIntent.getBroadcast(
                    registrar.context(),
                    0,
                    sentIntent,
                    PendingIntent.FLAG_UPDATE_CURRENT
             );

            sentIntents.add(sentPendingIntent);

            Intent deliveredIntent = new Intent("SMS_DELIVERED");
            deliveredIntent.putExtra("sentId", sentId);
            PendingIntent deliveredPendingIntent = PendingIntent.getBroadcast(
                    registrar.context(),
                    UUID.randomUUID().hashCode(),
                    deliveredIntent,
                    PendingIntent.FLAG_UPDATE_CURRENT
             );
            deliveredIntents.add(deliveredPendingIntent);
        }
        sms.sendMultipartTextMessage(address, null, parts, sentIntents, deliveredIntents);
        // sms.sendTextMessage(address, null, body, sentPendingIntent, sentPendingIntent);
        result.success(null);
    }
}

@TargetApi(Build.VERSION_CODES.DONUT)
class SmsSender implements MethodCallHandler {
    private final Registrar registrar;
    private final Permissions permissions;

    SmsSender(Registrar registrar) {
        this.registrar = registrar;
        permissions = new Permissions(registrar.activity());
    }

    @Override
    public void onMethodCall(MethodCall call, MethodChannel.Result result) {
        if (call.method.equals("sendSMS")) {
            String address = call.argument("address").toString();
            String body = call.argument("body").toString();
            int sentId = call.argument("sentId");
            Integer subId = call.argument("subId");
            if (address == null) {
                result.error("#02", "missing argument 'address'", null);
            } else if (body == null) {
                result.error("#02", "missing argument 'body'", null);
            } else {
                SmsSenderMethodHandler handler = new SmsSenderMethodHandler(registrar, result, address, body, sentId, subId);
                this.registrar.addRequestPermissionsResultListener(handler);
                handler.handle(this.permissions);
            }
        } else {
            result.notImplemented();
        }
    }
}
